### PR TITLE
Fixes #449 clear history

### LINKED ISSUE #449: Clear History
see https://github.com/Open-Systems-Pharmacology/PK-Sim/issues/911

### PR DESCRIPTION
**Pull request by msevestre**

### CODE REVIEW

File: src/OSPSuite.Assets/ApplicationIcons.cs [hunk 1]
```diff
@@ -478,6 +478,7 @@ public static class ApplicationIcons
       public static readonly ApplicationIcon SnapshotImport = AddNamedIcon(Icons.SnapshotImport, "SnapshotImport");
       public static readonly ApplicationIcon File = AddNamedIcon(Icons.ProjectNew, "File");
       public static readonly ApplicationIcon Redo = AddNamedIcon(Icons.Redo, "Redo");
+      public static readonly ApplicationIcon ClearHistory = AddNamedIcon(Icons.Delete, "ClearHistory");
```
Comment (msevestre) [inline src/OSPSuite.Assets/ApplicationIcons.cs]: We should probably get a new icon for that. For now this is the DELETE icon

File: src/OSPSuite.Core/Domain/Services/IHistoryManagerRetriever.cs [hunk 2]
```diff
@@ -0,0 +1,9 @@
+﻿using OSPSuite.Core.Commands.Core;
```
Comment (msevestre) [inline src/OSPSuite.Core/Domain/Services/IHistoryManagerRetriever.cs]: This service will have to be implemented in both apps, as the HistoryManager is not stored in the same place

File: src/OSPSuite.Infrastructure/Services/HistoryTask.cs [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+         _historyManagerRetriever = historyManagerRetriever;
+         _dialogCreator = dialogCreator;
+         _commandExecuter = commandExecuter;
+         _eventPublisher = eventPublisher;
+         _commandMetaDataRepository = commandMetaDataRepository;
+      }
+
+      public void ClearHistory()
+      {
+         var clear = _dialogCreator.MessageBoxYesNo(Captions.ReallyClearHistory, Captions.Clear, Captions.CancelButton);
```
Comment (msevestre) [inline src/OSPSuite.Infrastructure/Services/HistoryTask.cs]: Ask the user first

File: src/OSPSuite.Infrastructure/Services/HistoryTask.cs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+         if (clear == ViewResult.No) 
+            return;
+
+         clearProjectHistory();
+         clearCurrentHistory();
+      }
+
+      private void clearCurrentHistory()
```
Comment (msevestre) [inline src/OSPSuite.Infrastructure/Services/HistoryTask.cs]: This should clear the project

File: src/OSPSuite.Infrastructure/Services/HistoryTask.cs [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+      {
+         var historyManager = _historyManagerRetriever.Current;
```
Comment (msevestre) [inline src/OSPSuite.Infrastructure/Services/HistoryTask.cs]: This should clear all new commands created so far.
I will need to test all of that in PKSim to see if something is not working as expected

File: src/OSPSuite.Infrastructure/Services/HistoryTask.cs [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+         if (string.IsNullOrEmpty(projectPath))
+            return;
+
+         _commandExecuter.ExecuteCommand(projectPath, clearCommand);
+         _commandMetaDataRepository.Clear();
+      }
+
+      private void clearCommand(DbConnection connection)
+      {
+         connection.ExecuteNonQuery("DELETE FROM HISTORY_ITEMS;");
```
Comment (msevestre) [inline src/OSPSuite.Infrastructure/Services/HistoryTask.cs]: @Yuri05 These are the commands you gave me.

File: src/OSPSuite.Presentation/Presenters/ContextMenus/CommonMenuBarButtons.cs [hunk 7]
```diff
@@ -62,5 +60,14 @@ public static IMenuBarItem JournalDiagramView(MenuBarItemId menuBarItemId)
             .WithCommand<JournalDiagramVisibilityUICommand>()
             .WithIcon(ApplicationIcons.JournalDiagram);
       }
+
+      public static IMenuBarItem ClearHistory(MenuBarItemId menuBarItemId)
```
Comment (msevestre) [inline src/OSPSuite.Presentation/Presenters/ContextMenus/CommonMenuBarButtons.cs]: I created a button that can be used in ribbon as well as an entry in history presente.r That way we have the best of both worlds

**Comment by msevestre** [comment]
@Yuri05 Need to merge to move forward. Please review

File: src/OSPSuite.Presentation/Presenters/ContextMenus/CommonMenuBarButtons.cs (hunk 7 above)
Comment (Yuri05) [reply src/OSPSuite.Presentation/Presenters/ContextMenus/CommonMenuBarButtons.cs]: >... the best of both worlds

"Best of all worlds" always reminds me about [this](http://raybradbury.ru/library/story/60/4/0/)  :wink:

**Pull request merged**